src/Raven.Server/Documents/Sharding/Background/ShardedPeriodicDocumentsMigrator.cs

```diff
@@ -0,0 +1,98 @@
+        {
+            _database = database;
+        }
+
+        protected override async Task DoWork()
+        {
+            await WaitOrThrowOperationCanceled(_database.Configuration.Sharding.PeriodicDocumentsMigrationInterval.AsTimeSpan);
+
+            while (_database.ServerStore.Sharding.HasActiveMigrations(_database.ShardedDatabaseName))
+                await WaitOrThrowOperationCanceled(TimeSpan.FromMilliseconds(300));
+
+            await ExecuteMoveDocuments();
+        }
+
+        internal async Task ExecuteMoveDocuments()
+        {
+            try
+            {
+                int bucket = -1;
+                int moveToShard = -1;
+                bool found = false;
+                using (_database.ShardedDocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
+                using (context.OpenReadTransaction())
+                {
+                    var configuration = _database.ShardingConfiguration;
+                    for (var index = 0; index < configuration.BucketRanges.Count; index++)
+                    {
+                        var range = configuration.BucketRanges[index];
+                        if (range.ShardNumber == _database.ShardNumber)
+                            continue;
+
+                        var start = range.BucketRangeStart;
+                        var end = index == configuration.BucketRanges.Count - 1
+                            ? int.MaxValue
+                            : configuration.BucketRanges[index + 1].BucketRangeStart;
+
+                        var bucketStatistics = ShardedDocumentsStorage.GetBucketStatistics(context, start, end);
+
+                        if (bucketStatistics == null)
+                            continue;
+
+                        foreach (var bucketStats in bucketStatistics)
+                        {
+                            if (bucketStats.NumberOfDocuments == 0)
+                                continue;
+
+                            bucket = bucketStats.Bucket;
+                            moveToShard = range.ShardNumber;
+                            found = true;
+                            break;
+                        }
+
+                        if (found)
+                        {
+                            await MoveDocumentsToShard(bucket, moveToShard);
+                            return;
+                        }
+                    }
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                // this will stop processing
+                throw;
+            }
+            catch (Exception e)
+            {
+                if (Logger.IsOperationsEnabled)
+                    Logger.Operations($"Failed to execute documents migration for '{_database.Name}'", e);
+
+                throw;
+            }
+        }
+
+        private async Task MoveDocumentsToShard(int bucket, int moveToShard)
```

Comment:
If we have three buckets to move, what happens?
We start the migration, but what will handle the next one?

Reply:
But that would happen on the _next_ call, after we do another 60 sec wait, no?